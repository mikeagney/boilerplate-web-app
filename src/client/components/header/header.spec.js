import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Nav } from 'react-bootstrap';
import Routes from '../../pages/routes';
import Header from './header';

jest.mock('../../pages/routes');

describe('Header component', () => {
  const defaultProps = {
    location: {
      pathname: '/',
    },
    history: {
      push: jest.fn(),
    },
  };
  const navRoutes = [
    { key: 'route1', path: '/route1', name: 'Route 1' },
    { key: 'route2', path: '/route2', name: 'Route 2' },
  ];

  beforeEach(() => {
    jest.clearAllMocks();
    Routes.getNavRoutes.mockReturnValue(navRoutes);
  });

  it('will render the header component', () => {
    // Arrange
    const props = defaultProps;
    const wrapper = shallow(<Header {...props} />);

    // Act
    // Assert
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find(Nav.Item).length).toEqual(navRoutes.length);
    expect(Routes.getMatchingRoute).toHaveBeenCalledWith('/');
  });

  it('will set active nav based on the matched route', () => {
    // Arrange
    const props = defaultProps;
    const matchedKey = 'route2';
    Routes.getMatchingRoute.mockReturnValue({ key: matchedKey });
    const wrapper = shallow(<Header {...props} />);

    // Act
    // Assert
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find(Nav).prop('activeKey')).toEqual(matchedKey);
  });

  it('will do nothing when an event key for a nonexistent route is selected', () => {
    // Arrange
    const props = defaultProps;
    const wrapper = shallow(<Header {...props} />);
    const header = wrapper.instance();
    Routes.getNavRoute.mockReturnValue(undefined);

    // Act
    header.onSelect('route3');

    // Assert
    expect(Routes.getNavRoute).toHaveBeenCalledWith('route3');
    expect(defaultProps.history.push).not.toHaveBeenCalled();
  });

  it('will push to history when a route is selected', () => {
    // Arrange
    const props = defaultProps;
    const wrapper = shallow(<Header {...props} />);
    const header = wrapper.instance();
    Routes.getNavRoute.mockReturnValue({ path: '/foo' });

    // Act
    header.onSelect('route4');

    // Assert
    expect(Routes.getNavRoute).toHaveBeenCalledWith('route4');
    expect(defaultProps.history.push).toHaveBeenCalledWith('/foo');
  });
});
