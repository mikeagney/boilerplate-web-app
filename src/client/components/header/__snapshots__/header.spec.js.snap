// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component will render the header component 1`] = `
<div
  className="myapp-header"
>
  <Bootstrap(undefined)
    bg="dark"
    expand="sm"
    variant="dark"
  >
    <NavbarToggle
      aria-controls="header-nav"
      as="button"
      label="Toggle navigation"
    />
    <NavbarCollapse
      id="header-nav"
    >
      <Nav
        as="div"
        fill={false}
        justify={false}
        onSelect={[Function]}
      >
        <NavItem
          as="div"
          key="route1"
        >
          <NavLink
            as={[Function]}
            disabled={false}
            eventKey="route1"
          >
            Route 1
          </NavLink>
        </NavItem>
        <NavItem
          as="div"
          key="route2"
        >
          <NavLink
            as={[Function]}
            disabled={false}
            eventKey="route2"
          >
            Route 2
          </NavLink>
        </NavItem>
      </Nav>
    </NavbarCollapse>
  </Bootstrap(undefined)>
</div>
`;

exports[`Header component will set active nav based on the matched route 1`] = `
<div
  className="myapp-header"
>
  <Bootstrap(undefined)
    bg="dark"
    expand="sm"
    variant="dark"
  >
    <NavbarToggle
      aria-controls="header-nav"
      as="button"
      label="Toggle navigation"
    />
    <NavbarCollapse
      id="header-nav"
    >
      <Nav
        activeKey="route2"
        as="div"
        fill={false}
        justify={false}
        onSelect={[Function]}
      >
        <NavItem
          as="div"
          key="route1"
        >
          <NavLink
            as={[Function]}
            disabled={false}
            eventKey="route1"
          >
            Route 1
          </NavLink>
        </NavItem>
        <NavItem
          as="div"
          key="route2"
        >
          <NavLink
            as={[Function]}
            disabled={false}
            eventKey="route2"
          >
            Route 2
          </NavLink>
        </NavItem>
      </Nav>
    </NavbarCollapse>
  </Bootstrap(undefined)>
</div>
`;
